lib/Dialect/SPIRV/Transforms/DecorateSPIRVCompositeTypeLayoutPass.cpp

```diff
@@ -0,0 +1,318 @@
+  VulkanLayoutUtils::Alignment maxMemberAlignment = 1;
+  VulkanLayoutUtils::Alignment memberAlignment = 1;
+
+  for (uint32_t i = 0, e = structType.getNumElements(); i < e; ++i) {
+    auto memberType = VulkanLayoutUtils::decorateType(
+        structType.getElementType(i), memberSize, memberAlignment);
+    structMemberOffset = llvm::alignTo(structMemberOffset, memberAlignment);
+    memberTypes.push_back(memberType);
+    layoutInfo.push_back(structMemberOffset);
+    // According to the Vulkan spec:
+    // "A structure has a base alignment equal to the largest base alignment of
+    // any of its members."
+    structMemberOffset += memberSize;
+    maxMemberAlignment = std::max(maxMemberAlignment, memberAlignment);
+    memberAlignment = 1;
+    memberSize = 0;
+  }
+
+  size = llvm::alignTo(structMemberOffset, maxMemberAlignment);
+  alignment = maxMemberAlignment;
+  structType.getMemberDecorations(memberDecorations);
+  return spirv::StructType::get(memberTypes, layoutInfo, memberDecorations);
+}
+
+Type VulkanLayoutUtils::decorateType(
+    Type type, spirv::StructType::LayoutInfo &size,
+    VulkanLayoutUtils::Alignment &alignment) {
+  if (spirv::SPIRVDialect::isValidScalarType(type)) {
+    alignment = VulkanLayoutUtils::getScalarTypeAlignment(type);
+    // Vulkan spec does not specify any padding for a scalar type.
+    size = alignment;
+    return type;
+  }
+
+  switch (type.getKind()) {
+  case spirv::TypeKind::Struct:
+    return VulkanLayoutUtils::decorateType(type.cast<spirv::StructType>(),
+                                              size, alignment);
+  case spirv::TypeKind::Array:
+    return VulkanLayoutUtils::decorateType(type.cast<spirv::ArrayType>(),
+                                              size, alignment);
+  case StandardTypes::Vector:
+    return VulkanLayoutUtils::decorateType(type.cast<VectorType>(), size,
+                                              alignment);
+  default:
+    llvm_unreachable("unhandled SPIR-V type");
+  }
+}
+
+Type VulkanLayoutUtils::decorateType(
+    VectorType vectorType, spirv::StructType::LayoutInfo &size,
+    VulkanLayoutUtils::Alignment &alignment) {
+  const auto numElements = vectorType.getNumElements();
+  auto elementType = vectorType.getElementType();
+  spirv::StructType::LayoutInfo elementSize = 0;
+  VulkanLayoutUtils::Alignment elementAlignment = 1;
+
+  auto memberType = VulkanLayoutUtils::decorateType(elementType, elementSize,
+                                                       elementAlignment);
+  // According to the Vulkan spec:
+  // 1. "A two-component vector has a base alignment equal to twice its scalar
+  // alignment."
+  // 2. "A three- or four-component vector has a base alignment equal to four
+  // times its scalar alignment."
+  auto vectorTypeAlignment =
+      numElements == 2 ? elementAlignment * 2 : elementAlignment * 4;
+
+  auto vectorSize = elementSize * numElements;
+  size = llvm::alignTo(vectorSize, vectorTypeAlignment);
```

Comment:
Here too. This method is for vector type itself and return the real size for the vector. It should not worry about padding to match the alignment. That should be the responsibility of an outer type. (And the outer type may act differently regarding padding.)